python/rmm/_lib/device_buffer.pyx

```diff
@@ -278,18 +278,25 @@ cdef class DeviceBuffer:
 
         return b
 
-    cdef size_t c_size(self):
+    cdef size_t c_size(self) except *:
         return self.c_obj.get()[0].size()
 
-    cpdef void resize(self, size_t new_size):
+    cpdef void resize(self, size_t new_size) except *:
         self.c_obj.get()[0].resize(new_size)
 
-    cpdef size_t capacity(self):
+    cpdef size_t capacity(self) except *:
         return self.c_obj.get()[0].capacity()
 
-    cdef void* c_data(self):
+    cdef void* c_data(self) except *:
         return self.c_obj.get()[0].data()
```

Comment:
These are [marked `noexcept` in C++]( https://github.com/rapidsai/rmm/blob/bb7ff451b9ce3b98e7c9bdc148ffea40880144c4/include/rmm/device_buffer.hpp#L330-L360 ). So I don't think we need these.

Reply:
👍

Reply:
But wait -- we are doing more than just calling `self.c_obj.data()` for example. What if `c_obj.get()` throws, or the dereferencing. Those exceptions get swallowed.

Reply:
Well [`.get()` is also marked `noexcept`]( https://en.cppreference.com/w/cpp/memory/unique_ptr/get ) and `.data()` was already included in the `noexcept` above.

In any event, if there are C++ functions/methods that Cython exposes that do raise exceptions, these should be marked correctly in Cython (or fixed as needed) as opposed to handling them here.

Reply:
> In any event, if there are C++ functions/methods that Cython exposes that do raise exceptions, these should be marked correctly in Cython (or fixed as needed) as opposed to handling them here.

There are two kinds of `except` clauses to consider here: 

1. Exceptions thrown by C++: to correctly capture and re-raise these as Python `RuntimeError`s, we need to mark the wrapped functions as `except +`.

2. Python exceptions thrown in `cdef` functions: for `cdef` functions that don't return a Python object (like this one), any Python exceptions that may occur within them -- including the `RuntimeError`s mentioned in (1)  -- will be ignored, unless we mark those `cdef` functions `except *` or `except ?`.

(1) and (2) are somewhat independent, in that just doing one does not eliminate the need for the other. I agree that for C++ functions marked `noexcept`, we can safely omit (1).

But omitting (2) is dangerous and not future proof. It's too easy for an implementation of a `cdef` function to change to include calls to C++/Python code that can throw. Furthermore, even if all the code inside a `cdef` function is "guaranteed" not to throw, it's still possible for runtime errors to originate from inside that function, as I demonstrated here: https://github.com/pandas-dev/pandas/pull/33723